### PR TITLE
switch from many parameters in constructor to a single Options struct

### LINKED ISSUE #20: Make all parameters configurable
Currently the constructor accepts a [fixed number of arguments](https://github.com/mapbox/geojson-vt-cpp/blob/b2d18e2a4abd1deb0b0e6d3f5ed634bb53574fe7/include/mapbox/geojsonvt/geojsonvt.hpp#L22) for parameters, and all others (in particular `buffer`) are hardcoded.

We need to make every parameter configurable (like in JS). To keep the constructor sane, we should also switch to passing a map of options rather than a flat list.

This is pretty important because we want to bump default annotation tiles buffer in mapbox-gl-native. (related: https://github.com/mapbox/mapbox-gl-native/issues/1673)

Another issue is that mapbox-gl-native [got some parameters hardcoded](https://github.com/mapbox/mapbox-gl-native/commit/03c4ed75a2684b9866a00eebf0392c96cceb8d34) to values which are now way different in GeoJSON-VT.

cc @jfirebaugh @incanus @mikemorris

### PR DESCRIPTION
**Pull request by mikemorris**
**semver major**, fixes #20

/cc @mourner @jfirebaugh @incanus

### CODE REVIEW

**Comment by incanus** [comment]
:+1:, especially the abstraction of extent and buffer.

**Comment by mourner** [comment]
Looks great. 

I'd suggest mirroring the JS version API though, with features as the first flat required argument and optional object for all the rest options. 

BTW what's the reason to keep the API 2-step with first doing `convertFeatures` on GeoJSON data and then passing the converted features to constructor, instead of one step like in the JS version?

**Comment by mikemorris** [comment]
> I'd suggest mirroring the JS version API though, with features as the first flat required argument and optional object for all the rest options.

Sounds good, will change this. :+1: 

> BTW what's the reason to keep the API 2-step with first doing convertFeatures on GeoJSON data and then passing the converted features to constructor, instead of one step like in the JS version?

Not sure, I'm not too familiar with either version to be honest.

**Comment by mourner** [comment]
@incanus do you know?

**Comment by mikemorris** [comment]
@jfirebaugh Is there an easy fix for initializer lists in GCC?
```
../test/test_full.cpp: In function ‘std::map<std::basic_string<char>, std::vector<mapbox::util::geojsonvt::TileFeature> > genTiles(const string&, uint8_t, uint32_t)’:
../test/test_full.cpp:21:55: error: no matching function for call to ‘mapbox::util::geojsonvt::GeoJSONVT::GeoJSONVT(<brace-enclosed initializer list>)’
                         .indexMaxPoints = maxPoints } };
                                                       ^
../test/test_full.cpp:21:55: note: candidate is:
In file included from ../test/test_full.cpp:4:0:
../include/mapbox/geojsonvt/geojsonvt.hpp:33:5: note: mapbox::util::geojsonvt::GeoJSONVT::GeoJSONVT(mapbox::util::geojsonvt::Options)
     GeoJSONVT(Options);
     ^
../include/mapbox/geojsonvt/geojsonvt.hpp:33:5: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘mapbox::util::geojsonvt::Options’
```

**Comment by jfirebaugh** [comment]
That's [designated initializer](https://gcc.gnu.org/onlinedocs/gcc/Designated-Inits.html) syntax, and it's a C99 feature. As far as I know there's no portable way to use it in C++. You'll have to write it out manually:

```
    Options options;
    options.features = features;
    options.maxZoom = 14;
    options.indexMaxZoom = maxZoom;
    options.indexMaxPoints = maxPoints;
    GeoJSONVT index(options);
```

**Comment by incanus** [comment]
> BTW what's the reason to keep the API 2-step with first doing convertFeatures on GeoJSON data and then passing the converted features to constructor, instead of one step like in the JS version?

Hmm, interesting, and I didn't do it in the intermediary Swift port: 

https://github.com/mapbox/geojsonvt.swift/blob/bc44007a3baf7d26734e108e5d05079d037f39cd/geojsonvt/geojsonvt/geojsonvt.swift

I think this was related to something in the original shape annotations implementation (https://github.com/mapbox/mapbox-gl-native/pull/1655/files) like needing to convert first, but not necessarily tile until a later point.

**Comment by mourner** [comment]
> I think this was related to something in the original shape annotations implementation 

@incanus can we ditch the step now so that it mirrors the JS API?

**Comment by incanus** [comment]
Fine by me, particularly since this is now a separate project.

File: src/geojsonvt.cpp [hunk 1]
```diff
@@ -44,17 +44,14 @@ GeoJSONVT::convertFeatures(const std::string& data, uint8_t maxZoom, double tole
...
-                     uint8_t indexMaxZoom_,
-                     uint32_t indexMaxPoints_,
-                     bool solidChildren_,
-                     double tolerance_)
-    : maxZoom(maxZoom_),
-      indexMaxZoom(indexMaxZoom_),
-      indexMaxPoints(indexMaxPoints_),
-      solidChildren(solidChildren_),
-      tolerance(tolerance_) {
+GeoJSONVT::GeoJSONVT(std::vector<ProjectedFeature> features_, Options options)
```
Comment (kkaefer) [inline src/geojsonvt.cpp]: Why not store an `Options` object directly into the `GeoJSONVT` object rather than exploding it into its parts?

Comment (mikemorris) [inline src/geojsonvt.cpp]: Hmm that seems like it would work well.

File: include/mapbox/geojsonvt/geojsonvt.hpp [hunk 2]
```diff
@@ -13,18 +13,34 @@ namespace mapbox {
...
-                                                         double tolerance = 3);
+    GeoJSONVT(const std::string&, Options = Options());
 
-    GeoJSONVT(std::vector<ProjectedFeature> features_,
-              uint8_t maxZoom = 14,
-              uint8_t indexMaxZoom = 5,
-              uint32_t indexMaxPoints = 100000,
-              bool solidChildren = false,
-              double tolerance = 3);
+    std::vector<ProjectedFeature> convertFeatures(const std::string& data);
```
Comment (mikemorris) [inline include/mapbox/geojsonvt/geojsonvt.hpp]: Hmm, this should probably be a private method now...

**Comment by mikemorris** [comment]
@mourner @kkaefer Is it valuable to keep `convertFeatures` as a public static method for [performance testing](https://github.com/mapbox/geojson-vt-cpp/blob/25ab25807cc2f0ed58d7710f665861df925b7941/debug/debug.cpp#L92-L93)?

**Comment by mourner** [comment]
@mikemorris doesn't matter much, I think this is good to merge.

**Comment by kkaefer** [comment]
@mikemorris let's make it private; we should add more detailed benchmarking to the internals of geojson-vt-cpp anyway if the need arises to optimize.

**Comment by mikemorris** [comment]
Rebased and squashed, this should be good to merge (and tag `v3.0` - is breaking from JS versions an issue here @mourner?) once Travis passes.

**Comment by mourner** [comment]
> and tag v3.0 - is breaking from JS versions an issue here @mourner?

Not sure what's the best practice is, but worth noting that we diverged the API anyway even though the version was mirrored, so we probably have to break the version mirroring here, even if it's not wanted. @jfirebaugh what do you think?

**Comment by jfirebaugh** [comment]
I think it's fine to version js and cpp independently.

**Pull request merged**